### ISSUE TITLE
Mosquitto usage 99.7 CPU

### ISSUE THREAD

**Issue by IsabelManiega** [iegomez/mosquitto-go-auth#64]
Hi!,

I have a problem in my server (Debian GNU/Linux 10 (buster)), I put the plugin in the mosquitto configuration, at this moment It shows high cpu usage.

we have the plugin in another server and it is working, but the connection plugin-mosquitto has a delay of 10 seconds.

I do not know if the problem is with the mosquitto or with our own servers,
I have described the problem and the configurations as follows:

![image](https://user-images.githubusercontent.com/25103614/81904542-b5d4b980-95c3-11ea-933e-bc93e7e817b8.png)

My mosquitto configuration is:

mosquitto.conf:

pid_file /var/run/mosquitto.pid

persistence true
persistence_location /var/lib/mosquitto/

log_dest file /var/log/mosquitto/mosquitto.log

include_dir /etc/mosquitto/conf.d

/conf.d/mosquitto-go-auth:
auth_plugin /var/mosquitto-go-auth/go-auth.so
auth_opt_backends mongo
auth_opt_check_prefix false
allow_anonymous false
 
# debug
auth_opt_log_level debug
 
#Mongo: 
auth_opt_mongo_host localhost
auth_opt_mongo_port 21017
auth_opt_mongo_dbname mosquitto
auth_opt_mongo_username mosquitto

auth_opt_mongo_password mosquitto_auth
auth_opt_mongo_users users
auth_opt_mongo_acls acls

I use Mosquitto 1.6.0 Version and  protocol mqtt 3.1.1 

Thank you in advance for your attention.

**Comment by iegomez** [comment]
This issue could be caused by different things (mosquitto, the plugin, mongo, some other program, etc.), so I think the best would be to discard potential culprits by testing different scenarios:

- Does it run fine without the plugin?
- What if you run the plugin only the Files backend?
- What if you run it with other backends other than Mongo?
- What if you time how much it takes since the plugin gets a request and Mongo is done doing authentication/authorization by throwing some timers and logs in plugin?
- Etc.

**Comment by iegomez** [comment]
Hi, @IsabelManiega . Do you have any more details to share about those points?

**Comment by sanhardik** [comment]
Hello @iegomez 
I am facing the same issue. 
Mosquitto Version: 1.6.8, Protocol MQTT 3.1.1
With Plugin, CPU usage is 100%
Without Plugin, CPU usage is around 3%. 

The Plugin is using MySQL 
Without plugin, I am using basic password_file and acl_file.

**Comment by sanhardik** [comment]
We tried Auth Plugin v0.6.0, v0.5.0 and v0.4.0 but same results

**Comment by iegomez** [comment]
Hi, @sanhardik, and thanks for giving more info.

As I asked in the original report, could you try some alternative configurations so we can narrow this down a bit? For example, trying the plugin with the Files backend instead of no plugin with static files, or some other plugin such as Redis. I just wanna know if DB drivers might have something to do with it.

Also, it'd be interesting to see if running different versions of Mosquitto like 1.5.x or the latest 1.6.10 give the same results.

Finally, any more details on MySQL process(es) (cpu usage by process, active connections or whatever else you can gather) could be very helpful too.

Sadly, I haven't seen this happen so it's a bit hard to diagnose, so any more info could help us shed some light.

**Comment by sanhardik** [comment]
I changed it to use files backend and the CPU usage is 100%. 
One other observation: 
We have a service that publishes data to Broker.
Without Plugin, time taken per publish was around 1ms
With the Plugin, using Files backend, time taken per publish is around 26ms. 

I will try different mosquitto versions and report back. 

I am running logs for auth plugin with following options:
```
auth_opt_log_level debug
auth_opt_log_dest file
auth_opt_log_file /var/log/mosquitto/mosquitto.log
```
Are there any logs that I can enable to help us out with this ?

I am also not sure if Debian vs Ubuntu is causing any issues here

**Comment by iegomez** [comment]
That should give you all plugin related logs. For `mosquitto` you may set it to log everything:
```
log_type all
```
Alternatively, if you run it manually I think you can give it the verbose flag `-v`.

I'm inclined to say Debian vs Ubuntu shouldn't be a problem: I originally started this on Debian 9, but I've been using Linux Mint for a long time and haven't had any issues.

**Comment by iegomez** [comment]
> Without Plugin, time taken per publish was around 1ms
> With the Plugin, using Files backend, time taken per publish is around 26ms.

About that, it may vary. I'm not entirely sure and some reading would be needed, but I'd expect there to be a penalty just by calling a shared object, which is how an external plugin is loaded, instead of directly depending on mosquitto static files auth internals. That could be checked by trying with https://github.com/jpmens/mosquitto-auth-plug to see how much a pure C plugin overhead is.

Now, considering both that plugin and this one store all password hashes and acls in memory, the biggest difference should come from the use of `CGo`, which does have a severely greater impact than that of other languages FFIs. Granted, it's not normally a 24 ms difference, but adding the previous reason and whatever is causing the high CPU usage, it doesn't sound so crazy.

**Comment by sanhardik** [comment]
When I tried with Mosquitto v1.5.9, it wont run. It says 
```Error: Incorrect auth plugin version (got 4, expected 3).```

Auth Plugin Commit: 5b603bf

**Comment by iegomez** [comment]
I've seen that before when having Mosquittto 1.6.x installed and then downgrading to 1.5.x or viceversa, as well as switching between 1.4.x and 1.5.x. If I recall correctly, they bump `MOSQ_AUTH_PLUGIN_VERSION` between mosquitto's minor versions or whenever they introduce a breaking change for plugins (e.g. https://github.com/iegomez/mosquitto-go-auth/blob/master/auth-plugin.c#L8), but for some reason it was damn difficult to make mosquitto pick up the new plugin version when upgrading/downgrading and I can't remember how I dealt with that. Probably the easiest way to try different versions is using available Mosquitto Docker images so you don run into this issue.

**Comment by sanhardik** [comment]
@iegomez 
We created a new machine with Mosquitto v1.5.9 and saw the same CPU usage. 

I tried jpmens C plugin and it improved the performance significantly. Seeing much less CPU usage. 
So as per your suggestion,  it seems its something to do with GO. 
What help can I provide if we want to sort this out ?

**Comment by iegomez** [comment]
@sanhardik That's a bummer, I was really hoping it had to do with the mosquitto version since I've never experienced high CPU usage, while yours and @IsabelManiega 's are the only related issues ones that have been reported.

I still don't think it has to do with Go per se, I was just saying that this plugin will perform worse than a C one without a doubt, not only because C is generally faster but also because of the FFI calls which would have an overhead in any case, but worse with CGo being notably bad at that compared to other languages. But when I say bad I mean an overhead of anything between a couple hundred ns when calling C from Go, to a couple ms at worst when calling Go from C, so usually not a deal breaker when your bottleneck is an HTTP roundtrip, a DB connection, etc. But none of this would mean high CPU usage in a normal situation, so there must be something else putting pressure on the CPU.

So again sorry for asking you to do so much work to diagnose the situation, but any details on the environment, what else is running along with mosquitto and the plugin, and anything else you can think of, would allow me to run some stress tests and see if I can finally reproduce the scenario you're describing.

For now all I can say is I'm terribly sorry you're having a bad experience with the plugin and I'll do anything I can to try to fix it, I'm just struggling to find time to put the effort needed these days. My first idea on a temporal patch would be to see if running a docker image (thanks to @LolHens we now have an updated image with latest versions of mosquitto and the plugin at https://hub.docker.com/r/iegomez/mosquitto-go-auth) shows any improvement (which would hint there's something else messing with it), and if not, for now you may be better off running jpmen's plugin with an older version of mosquitto (I think it supported up to latest 1.4.x versions of mosquitto before it was archived, but that may have changed) until we're able to resolve it.

**Comment by sanhardik** [comment]
@iegomez There is no need to be sorry. I think we should be grateful for your work and getting the Auth Plugin which otherwise would have taken a long time for us to build. 

Its an AWS t2.large EC2 machine running Ubuntu 16.04.
Its not running anything apart from the mosquito service and MySQL server for auth. 

We push in a lot of data (think of 1000 of messages a second). 
We are trying to run jpmen's plugin with latest mosquito version just to check if it will work.

**Comment by iegomez** [comment]
Ah, well, it might be it's not coping well with so many messages. I think I opened a free AWS account some time ago, I'll run some tests on it and see where that gets us.

With jpmen's plugin you should have some issues like plugin version 4 not being supported: https://github.com/iegomez/mosquitto-go-auth/blob/master/auth-plugin.c#L89
I think `MOSQ_ACL_SUBSCRIBE` could be an issue too. But those should not be too hard to fix, so good luck for now!

**Comment by sanhardik** [comment]
We tried jpmen's plugin but since it doesnt support newer versions of Mosquitto, we are having a tough time getting it to work as it is. 

Did you get a chance to check on CPU usage issues ?
We did notice that the CPU usage issues came up with TLS turned on. Dont know if that helps. 

Whats the libwebsockets version you recommend or are using ?

**Comment by iegomez** [comment]
Sorry, I haven't got the chance to do a proper check. I'm not sure on `libwebsockets` version since I'm on my work machine right now, but I'll report it later. In any case, your findings on TLS look promising, I'll try to test that as soon as I get some time.

**Comment by iegomez** [comment]
Quick update: last development has been done on a fairly new machine and I never got to check if TLS was working, so actually there's no specific `libwebsockets` version I can report or recommend right now. 

I remember there were `mosquitto` issues with some specific websockets version a couple of years ago, but I have no record of what version that was since most of that info I kept on a work machine when developing the plugin for work related stuff, where I had proper use cases and infrastructure to test things. But nowadays I mostly rely on the test suite to add features and fix bugs, and don't get to try it in real environments, so that's probably pretty outdated and unhelpful :disappointed:.

On a related note, I recall I used and had to tweak a couple of projects intended for stressing a broker, but again, haven't touched them in quite some time. So if you can recommend anything to simulate your workload and see what's going on, that'd be awesome!

Let's end with a happy message: I already opened a PR for another issue and have a plan for a feature that has been waiting for a while, all that while Monday is a holiday here, so I might get the time to actually deploy a basic setup to AWS and see how that goes during the weekend. No promises, but a chance at least :smile:.

Cheers!

**Comment by iegomez** [comment]
@sanhardik I did an experiment on a $10 Linode instance (Ubuntu 19.04, Linode 2GB: 1 CPU, 50GB Storage, 2GB RAM) running just a couple of static sites with 0 traffic (so almost idle) and a fresh installation of mosquitto + plugin.

I wrote a quick local program that runs one instance of `mosquitto_sub` and 500 instances of `mosquitto_pub` using TLS concurrently and found some interesting things:

1) With the plugin's local cache (not the Redis one) enabled there's almost no difference between running the plugin with the `Files` backend versus running `mosquitto` **without auth, that is allowing any user.** Now, I did use the same user for each client, so it gets cached immediately and then all other hit the cache.
2) So then I tried disabling the plugin's cache forcing it to do a regular check and it bumps up max and average times 10x, but it again shows almost exactly the same max and average times as running `mosquitto` with static files check, which sees the same 10x increase in times.
3) CPU never gets to terrible usage, at most I saw 13% without the plugin, but with or without it enabled usage was usually around 7-8%.

In summary: I don't see the behaviour you're reporting and don really know what else may be done about it. Also, I'm pretty sure the 10x in times for both `plugin with cache => plugin with files and no cache` and `mosquitto without auth => mosquitto with static files auth` is dominated by password hashing and so you see deteriorated service in both cases but with almost exactly the same times.

So I'm :man_shrugging: and would appreciate any new ideas on how to test this further.

**Comment by iegomez** [comment]
Holy shit, I just realized I wasn't hashing the cached keys! That's bad. :disappointed: It won't change the issue, but I'll get it fixed.

**Comment by sanhardik** [comment]
Would you mind testing this on Ubuntu 16.04. 
In our case, we publish more data than subscribe. Can you turn your test script around to publish more messages ? 

We had a high CPU even if only publishing and no subscribers. 

What version of Mosquitto and Libwebsockets did you use ?

**Comment by iegomez** [comment]
> Would you mind testing this on Ubuntu 16.04.

That's gonna be a bit tricky since: 1) I only have that individual machine available and it's running live sites; 2) I'm not sure they offer 16.04 in Linode any more. 3) If it had to do with the version, I don't really have the time capacity to support older systems.

> In our case, we publish more data than subscribe. Can you turn your test script around to publish more messages ?
> 
> We had a high CPU even if only publishing and no subscribers.

I think you misread or I'm doing it right now. Maybe you mean more messages per publisher? Because I just fired one subscriber and 500 publishers, each one publishing one message to the same topic the subscriber was listening on.

> What version of Mosquitto and Libwebsockets did you use ?

That would be latest (1.6.10) for Mosquitto and 2.0.3-3build1 for libwebsockets.

**Comment by sanhardik** [comment]
Thanks for testing. We will try it on Ubuntu 18 with your configurations and report back.

**Comment by sanhardik** [comment]
@iegomez 
Any reason why you are not using the latest version of libwebsockets v4.x.x (https://github.com/warmcat/libwebsockets)
We are getting some issues with openssl while trying to compile libwebsockets.

Do you mind sharing the openssl version as well ?

**Comment by iegomez** [comment]
No particular reason, it's just the version available in my server packages. The version for `openssl` is `1.1.1c-1ubuntu4.1`.

If it helps, just yesterday I messed an upgrade of Linux Mint in my local machine and installed Mint 20 instead, so today I configured everything again and was able to build mosquitto and the plugin with versions `3.2.1-3` and `1.1.1f-1ubuntu2` for `libwebsockets` and `openssl` respectively.

**Comment by sanhardik** [comment]
@iegomez 
It works fine now after we upgraded 
We tested with the following versions
Ubuntu 18.04
mosquitto version 1.6.10
go version go1.14.2
libwebsockets v4.0
Commit Version: https://github.com/iegomez/mosquitto-go-auth/commit/7dd3f5a895d8fb20def448eca9a353806b62419a

Thanks for the help. 

Feel free to close the ticket.

**Comment by iegomez** [comment]
I'm very glad you were able to solve it! Also glad to know it had to do with older versions.

Thanks for your patience!
